### PR TITLE
Try to fix performance issues and add performance metrics

### PR DESCRIPTION
**Pull request by jpambrun**
This is a tentative fix, there are probably some tweaking still needed.

The first commit change the image.lut typedArray that is polluted by fields invert, voiLUT, windowWidth, etc into an object cachedLut containing the array, lutArray, along with those fields.

I renamed lut to cachedLut since I find it confusing as the lut is applied on a viewport basis, not an image basis. I am open to new suggestions. However, this is basically a hack so that we don't have to compute identical luts multiple time if the same lut is applied to multiple viewports. It would be best to create a singleton lut manager that cache multiple luts.

I have also changed lut arrays from Int16Array into Uint8ClampedArray and removed the now useless manual clamping code.

I also have streamlined the storedColorPixelDataToCanvasImageData code a bit.

The second commit adds some performance metrics to the image objects under stats. I'm not sure I like to litter the code like that for this purpose. Also, I have used performance.now() instead of Date.now() as it provides sub-millisecond accuracy which is required when a lot of the timing is in the 2-3 ms range. However, I think it is not available in IE9 (do we care?).

This ties in with #60 and #61, maybe they can be closed after this is integrated.

### CODE REVIEW

File: dist/cornerstone.js [hunk 1]
```diff
@@ -2213,13 +2219,10 @@ if(typeof cornerstone === 'undefined'){
...
     {
-      modalityLutValue = mlutfn(storedValue);
-      voiLutValue = vlutfn(modalityLutValue);
-      clampedValue = Math.min(Math.max(voiLutValue, 0), 255);
       if(!invert) {
-        lut[storedValue+ (-offset)] = Math.round(clampedValue);
+        lut[storedValue+ (-offset)] = vlutfn(mlutfn(storedValue));
       } else {
-        lut[storedValue + (-offset)] = Math.round(255 - clampedValue);
+        lut[storedValue + (-offset)] = vlutfn(mlutfn(storedValue));
```
Comment (swederik) [inline dist/cornerstone.js]: I think you are missing the (255 - x) here. Right now it appears both paths of the if statement are identical.

File: src/internal/storedColorPixelDataToCanvasImageData.js [hunk 2]
```diff
@@ -7,33 +7,38 @@
 
     function storedColorPixelDataToCanvasImageData(image, lut, canvasImageDataData)
     {
+
+        var start = performance.now();
+        var pixelData = image.getPixelData();
+        image.stats.lastGetPixelDataTime = performance.now() - start;
+
+        start = performance.now();
+        var minPixelValue = image.minPixelValue;
```
Comment (swederik) [inline src/internal/storedColorPixelDataToCanvasImageData.js]: Looks like there is a copy/paste duplicate declaration for minPixelValue.

**Comment by jpambrun** [comment]
Will look into that. I initially made the changes and tested it into my
concatenated cornerstone.js then tried to backport it using the diff..
Looks like a very bad workflow.

I have some other minor tweaks as well.

File: src/internal/storedPixelDataToCanvasImageData.js [hunk 3]
```diff
@@ -21,28 +21,51 @@
...
 
+        start = (window.performance ? performance.now() : Data.now());
         // NOTE: As of Nov 2014, most javascript engines have lower performance when indexing negative indexes.
         // We have a special code path for this case that improves performance.  Thanks to @jpambrun for this enhancement
-        if(minPixelValue < 0){
-            while(storedPixelDataIndex < localNumPixels) {
-                localCanvasImageDataData[canvasImageDataIndex] = localLut[localPixelData[storedPixelDataIndex++] + (-minPixelValue)]; // alpha
+
+        // Added two paths (Int16Array, Uint16Array) to avoid polymorphic deoptimization in chrome.
+        if (pixelData instanceof Int16Array) {
```
Comment (jpambrun) [inline src/internal/storedPixelDataToCanvasImageData.js]: I have added this fix as well, it prevents a significant slowdown when this function becomes polymorphic after being called on a Uint16 followed by a Int16 image (or vice versa). But it looks pretty stupid..

**Comment by swederik** [comment]
Merged, thanks!

**Pull request closed without merging**